nemo/collections/nlp/data/text_normalization/decoder_dataset.py

```diff
@@ -0,0 +1,151 @@
+        s_left  = s_words[max(0,start_idx-ctx_size):start_idx]
+        s_right = s_words[end_idx:end_idx+ctx_size]
+
+        # Process sil words and self words
+        for jx in range(len(s_left)):
+            if s_left[jx] == constants.SIL_WORD: s_left[jx] = ''
+            if s_left[jx] == constants.SELF_WORD: s_left[jx] = w_left[jx]
+        for jx in range(len(s_right)):
+            if s_right[jx] == constants.SIL_WORD: s_right[jx] = ''
+            if s_right[jx] == constants.SELF_WORD: s_right[jx] = w_right[jx]
+        for jx in range(len(c_s_words)):
+            if c_s_words[jx] == constants.SIL_WORD:
+                c_s_words[jx] = ''
+                if inst_dir == constants.INST_BACKWARD: c_w_words[jx] = ''
+            if c_s_words[jx] == constants.SELF_WORD: c_s_words[jx] = c_w_words[jx]
+
+        # Extract input_words and output_words
+        c_w_words = word_tokenize(' '.join(c_w_words))
+        c_s_words = word_tokenize(' '.join(c_s_words))
+        w_input = w_left + [extra_id_0] + c_w_words + [extra_id_1] + w_right
+        s_input = s_left + [extra_id_0] + c_s_words + [extra_id_1] + s_right
+        if inst_dir == constants.INST_BACKWARD:
+            input_words = [constants.ITN_PREFIX] + s_input
+            output_words = c_w_words
+        if inst_dir == constants.INST_FORWARD:
+            input_words = [constants.TN_PREFIX] + w_input
+            output_words = c_s_words
+        # Finalize
+        self.input_str = ' '.join(input_words)
+        self.output_str = ' '.join(output_words)
+        self.direction = inst_dir
+        self.semiotic_class = semiotic_class
+
+class TextNormalizationDecoderDataset:
+    def __init__(
+            self,
+            input_file: str,
+            tokenizer: PreTrainedTokenizerBase,
+            mode: str,
+            max_len: int,
+            decoder_data_augmentation: bool
+        ):
+        assert(mode in constants.MODES)
+        self.mode = mode
+        raw_insts = read_data_file(input_file)
+
+        # Convert raw instances to TaggerDataInstance
+        insts, inputs, targets = [], [], []
+        for (classes, w_words, s_words) in tqdm(raw_insts):
+            for ix, (_class, w_word, s_word) in enumerate(zip(classes, w_words, s_words)):
+                if s_word in constants.SPECIAL_WORDS: continue
+                for inst_dir in constants.INST_DIRECTIONS:
+                    if inst_dir == constants.INST_BACKWARD and mode == constants.TN_MODE: continue
+                    if inst_dir == constants.INST_FORWARD and mode == constants.ITN_MODE: continue
+                    # Create a DecoderDataInstance
+                    inst = DecoderDataInstance(w_words, s_words, inst_dir,
+                                               start_idx=ix, end_idx=ix+1,
+                                               semiotic_class=_class)
+                    insts.append(inst)
+                    if decoder_data_augmentation:
```

Comment:
is this important?

Reply:
Yes. I found doing some simple data augmentation can improve the models' robustness. I will write the details in the rst doc.